File: Source/DotNET/Backend/Config/Configuration.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;

namespace Dolittle.Vanir.Backend.Config
{
    public class Configuration
    {
        public string RouteSegment { get; set; } = "";
        public bool IsRooted { get; set; } = false;
        public int Port { get; set; } = 80;
        public Guid MicroserviceId { get; set; } = Guid.Empty;
        public DatabaseConfiguration Database { get; set; } = new DatabaseConfiguration();
        public DatabaseConfiguration EventStore { get; set; } = new DatabaseConfiguration();
        public DolittleConfiguration Dolittle { get; set; } = new DolittleConfiguration();
        public string Environment { get; set; } = "development";

        public bool IsDevelopment => Environment == "development";
        public bool IsProduction => Environment == "production";

        public string Prefix => $"{(IsRooted ? "" : "/_")}{(RouteSegment.Length > 0 ? "/" : "")}{RouteSegment}";
        public string GraphQLRoute => $"{Prefix}/graphql";

        public string GraphQLPlaygroundRoute => $"{GraphQLRoute}/ui";
    }
}

File: Source/typescript/backend/Configuration.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import nconf from 'nconf';
import defaults from './DefaultConfiguration';
import { DolittleConfiguration } from './DolittleConfiguration';


export class Configuration {
    routeSegment = '';
    isRooted = false;
    publicPath = '';
    port = 80;
    microserviceId = '';

    dolittle: DolittleConfiguration;

    environment = 'development';

    constructor() {
        this.dolittle = new DolittleConfiguration();
    }

    get isDevelopment() {
        return this.environment === 'development';
    }

    get isProduction() {
        return this.environment === 'production';
    }

    static create(): Configuration {
        nconf
            .argv({ parseValues: true })
            .env({ separator: '__', parseValues: true, lowerCase: true })
            .file({ file: 'vanir.json' })
            .defaults(defaults);

        const instance = new Configuration();
        const source = nconf.get();

        for (const property in instance) {
            if (source[property]) {
                instance[property] = source[property];
            }
        }

        instance.environment = nconf.get('node:env') || 'development';

        return instance;
    }
}
